File: lib/services/acronym_access_service.dart
import 'package:le_dariel/models/acronym.dart';
import 'package:le_dariel/services/auth_service.dart';
import 'package:le_dariel/services/acronym_manager.dart';

class AcronymAccessService {
  final AuthService authService;
  final AcronymManager acronymManager;

  AcronymAccessService({
    required this.authService,
    required this.acronymManager,
  });

  List<Acronym> getAccessibleAcronyms() {
    final user = authService.currentUser;
    if (user == null) return [];

    if (user.canAccessAllAcronyms) {
      return acronymManager.acronyms;
    } else {
      // Utilisateur gratuit : limiter à 20 acronymes
      return acronymManager.acronyms.take(20).toList();
    }
  }

  bool canAccessAcronym(Acronym acronym) {
    final user = authService.currentUser;
    if (user == null) return false;

    if (user.canAccessAllAcronyms) {
      return true;
    } else {
      final accessibleAcronyms = getAccessibleAcronyms();
      return accessibleAcronyms.any((a) => a.id == acronym.id);
    }
  }

  int getLockedAcronymsCount() {
    final user = authService.currentUser;
    if (user == null) return acronymManager.acronyms.length;

    if (user.canAccessAllAcronyms) {
      return 0;
    } else {
      return (acronymManager.acronyms.length - 20).clamp(0, double.maxFinite.toInt());
    }
  }

  List<Acronym> searchAccessibleAcronyms(String query) {
    final accessibleAcronyms = getAccessibleAcronyms();

    if (query.isEmpty) {
      return accessibleAcronyms;
    }

    final lowerQuery = query.toLowerCase();
    return accessibleAcronyms.where((acronym) {
      return acronym.letters.toLowerCase().contains(lowerQuery) ||
          acronym.fullForm.toLowerCase().contains(lowerQuery);
    }).toList();
  }

  List<String> getSuggestions(String query) {
    final accessibleAcronyms = getAccessibleAcronyms();
    final lowerQuery = query.toLowerCase();
    
    final matchingAcronyms = accessibleAcronyms.where((acronym) {
      return acronym.letters.toLowerCase().contains(lowerQuery);
    }).toList();
    
    return matchingAcronyms.map((a) => a.letters).toList();
  }
}

File: Le_DARIEL/Services/AcronymAccessService.swift
//
//  AcronymAccessService.swift
//  Le_DARIEL
//
//  Created by FAEI971 on 23/07/2025.
//

import Foundation

class AcronymAccessService: ObservableObject {
    let authService: AuthService
    let acronymManager: AcronymManager
    
    init(authService: AuthService, acronymManager: AcronymManager) {
        self.authService = authService
        self.acronymManager = acronymManager
    }
    
    // MARK: - Access Control
    
    func getAccessibleAcronyms() -> [Acronym] {
        guard let user = authService.currentUser else {
            return []
        }
        
        if user.canAccessAllAcronyms {
            return acronymManager.acronyms
        } else {
            // Utilisateur gratuit : limiter à 20 acronymes
            return Array(acronymManager.acronyms.prefix(20))
        }
    }
    
    func canAccessAcronym(_ acronym: Acronym) -> Bool {
        guard let user = authService.currentUser else {
            return false
        }
        
        if user.canAccessAllAcronyms {
            return true
        } else {
            // Vérifier si l'acronyme fait partie des 20 premiers
            let accessibleAcronyms = getAccessibleAcronyms()
            return accessibleAcronyms.contains { $0.id == acronym.id }
        }
    }
    
    func getLockedAcronymsCount() -> Int {
        guard let user = authService.currentUser else {
            return acronymManager.acronyms.count
        }
        
        if user.canAccessAllAcronyms {
            return 0
        } else {
            return max(0, acronymManager.acronyms.count - 20)
        }
    }
    
    func searchAccessibleAcronyms(query: String) -> [Acronym] {
        let accessibleAcronyms = getAccessibleAcronyms()
        
        guard !query.isEmpty else {
            return accessibleAcronyms
        }
        
        return accessibleAcronyms.filter {
            $0.letters.localizedCaseInsensitiveContains(query) ||
            $0.fullForm.localizedCaseInsensitiveContains(query)
        }
    }
    
    func getSuggestions(for query: String) -> [String] {
        let accessibleAcronyms = getAccessibleAcronyms()
        let matchingAcronyms = accessibleAcronyms.filter {
            $0.letters.localizedCaseInsensitiveContains(query)
        }
        return matchingAcronyms.map { $0.letters }
    }
}

